### DISCUSSION TITLE
posts/left-closed-right-open

### DISCUSSION THREAD

**Discussion by giscus** [EarthMessenger/blog.earthmessenger.xyz#11] [giscus]
# 2023/02/13/left-closed-right-open

前言

http://blog.earthmessenger.xyz/2023/02/13/left-closed-right-open.html

**Comment by Menci** [comment]
二分用左闭右开的话 `l = mid + 1` 就不需要 `+ 1` 了吧

**Reply by EarthMessenger** [reply]
这里已经判断了 mid 是否为可行的答案，以后不需要再判断，所以把 mid 排除在区间外了。

**Reply by Menci** [reply]
对，我是指左闭右开的二分不需要 `mid + 1` 也不影响算法中止，不影响正确，不是你说的那样没用。因为左闭右闭的写法如果向下取整的话这里是必须 `mid + 1` 的，否则不会中止。

**Reply by EarthMessenger** [reply]
比如说 `l = 0, r = 1`，则 `mid = 0`，此时如果判断 `mid` 不是可行答案的话 `l` 必须等于 `mid + 1`，否则不会中止。

不过有可能我这份代码不够优，应该有别的更优美的写法。

**Reply by Menci** [reply]
左闭右开的 `l = 0, r = 1` 是结束

**Reply by EarthMessenger** [reply]
我尝试修改了下 `lower_bound` 的板子：

```cpp
int l = 0, r = n;
if (a[0] >= 0) return 0;
while (r - l > 1) {
    int mid = (l + r) / 2;
    if (a[mid] >= x) {
        r = mid;
    } else {
        l = mid;
    }
}
return r;
```

如果以 `r - l > 1` 作为循环条件的话 `mid` 永远取不到 0，需要特判，感觉不如原代码优美。有没有什么更好的实现啊？

**Comment by Enonya** [comment]
左闭右开邪教，感觉不如闭区间。

**Reply by fat-old-eight** [reply]
正确的
